server/src/main/java/org/apache/iotdb/db/tools/watermark/GroupedLSBWatermarkEncoder.java

```diff
@@ -0,0 +1,145 @@
+import java.util.List;
+import org.apache.iotdb.db.conf.IoTDBConfig;
+import org.apache.iotdb.tsfile.file.metadata.enums.TSDataType;
+import org.apache.iotdb.tsfile.read.common.Field;
+import org.apache.iotdb.tsfile.read.common.RowRecord;
+import org.apache.thrift.EncodingUtils;
+
+public class GroupedLSBWatermarkEncoder implements WatermarkEncoder {
+
+  private String secretKey;
+  private String bitString;
+  private int markRate = 2;
+  private int groupNumber;
+  private int maxBitPosition = 5;
+  private int minBitPosition = 0;
+
+  public GroupedLSBWatermarkEncoder(String secretKey, String bitString) {
+    this.secretKey = secretKey;
+    this.bitString = bitString;
+    this.groupNumber = bitString.length();
+  }
+
+  public GroupedLSBWatermarkEncoder(IoTDBConfig conf) {
+    this(conf.getWatermarkSecretKey(), conf.getWatermarkBitString());
+    this.markRate = conf.getWatermarkParamMarkRate();
+    this.maxBitPosition = conf.getWatermarkParamMaxRightBit();
+  }
+
+  public static int hashMod(String val, Integer base) {
+    MessageDigest md;
+    try {
+      md = MessageDigest.getInstance("MD5");
+    } catch (NoSuchAlgorithmException e) {
+      e.printStackTrace();
+      throw new RuntimeException("ERROR: Cannot find MD5 algorithm!");
+    }
+    md.update(val.getBytes());
+    BigInteger resultInteger = new BigInteger(1, md.digest());
+    return resultInteger.mod(new BigInteger(base.toString())).intValue();
+  }
+
+  private boolean getMarkFlag(long timestamp) {
+    return hashMod(String.format("%s%d", secretKey, timestamp), markRate) == 0;
+  }
+
+  private int getGroupId(long timestamp) {
+    return hashMod(String.format("%d%s", timestamp, secretKey), groupNumber);
+  }
+
+  private int getBitPosition(long timestamp) {
+    if (maxBitPosition <= minBitPosition) {
+      throw new RuntimeException("Error: minBitPosition is bigger than maxBitPosition");
+    }
+    int range = maxBitPosition - minBitPosition;
+    return minBitPosition + hashMod(String.format("%s%d%s", secretKey, timestamp, secretKey),
+        range);
+  }
+
+  private boolean getBitValue(long timestamp) {
+    int groupId = getGroupId(timestamp);
+    int bitIndex = groupId % bitString.length();
+    return bitString.charAt(bitIndex) == '1';
+  }
+
+  private int encodeInt(int value, long timestamp) {
+    int targetBitPosition = getBitPosition(timestamp);
+    boolean targetBitValue = getBitValue(timestamp);
+    return EncodingUtils.setBit(value, targetBitPosition, targetBitValue);
+  }
+
+  private long encodeLong(long value, long timestamp) {
+    int targetBitPosition = getBitPosition(timestamp);
+    boolean targetBitValue = getBitValue(timestamp);
+    return EncodingUtils.setBit(value, targetBitPosition, targetBitValue);
+  }
+
+  private float encodeFloat(float value, long timestamp) {
+    int integerPart = (int) value;
+    float decimalPart = value - integerPart;
+    return encodeInt(integerPart, timestamp) + decimalPart;
```

Comment:
Encoding integer part only may cause great loss when the number is small (especially for numbers like 0.251).
How about this, you know, referring to IEEE 754, you may turn it into an integer without changing its bits (using BitUtils if I do not get it wrong), and only encode the last fraction bits, which may minimize the change of the value.

Reply:
excellent